### ISSUE TITLE
2.0 Slow Timeline Load At Runtime

### ISSUE THREAD

**Issue by thebardsrc** [coppolaemilio/dialogic#1091]
findings of the slowdowns and where I left off on the first speedup PR:

main slowdowns is in the actual timeline loader. each line of the timeline takes ~80 ms to go through each pass of the "while idx < len(lines)-1:" loop, which adds up quickly, with the timeline we're basing this on having 110 lines, adds up to a little over 8 seconds of load time, front loaded at the beginning

characters are loaded as it reads the timeline, if that line has a character in it. character loading is only adding about 10 ms, so _that_ is happening pretty quickly.  

rest of the routines break down like this, based on time outputs for each further loop:
![image](https://user-images.githubusercontent.com/2214793/183007328-2749dcdd-a88f-4b9a-931e-a9fe8745b304.png)
Line 62 -> Line 69: ~5 ms
Line 69 -> Line 73: ~25 ms
LIne 73 -> Line 88: ~5 ms 
Line 88 -> Line 62 of next loop: ~45 ms

the actual DialogicUtil.get_event_by_string() loader is taking some of the time, but its not the biggest part of the time, that's those last four lines. They break down as follows:

event_content.replace(): 2ms 
event._load_from_string(): all the rest of this time 
events.append(): 3ms
prev_was_opener = event.can_contain_events(): 3ms

it's late now and I need to sleep and my ability to comprehend now is waning, so I can't go further into it.  but looking at  DialogicUtil.get_event_by_string() it's constantly reloading files every single loop, which really should be loaded only once and referenced from there (since all of DialogicUtil is a static class file, for reference this was the same problem DTS had in 1.x to cause it's extensive slowdown that I had to rewrite it to be not a static class for). and the actual event processing in event._load_from_string() should be held off to runtime instead, running that with each new line as its being processed rather than runnig that all at the beginning first

**Comment by thebardsrc** [comment]
I added a deferral commit to #1090 to check if it's in editor or not and defer event._load_from_string() to each event at runtime if its not. that speeds up the load-to-running time significantly, from 8 seconds to 2 seconds, and the load_from_screen() being called on each individual event just as its processed is imperceptible

because of how event._load_from_string() works, since it matches each one to an individual event, it still runs that so there is still some of that other load. I didn't look, but that _also_ might be able to be deferred as well, but then this whole loop would have to be rewritten to accomodate it

this test time doesn't have any jumps to any other timelines, either. so i have not looked at the recursive search yet because that slows things down even more, recursively stepping through each timeline and loading the characters and timleines referenced there. that doesnt actually need to happen at runtime, beyond verifiying that the jumps out of that timeline are valid files before trying to load them

**Comment by thebardsrc** [comment]
I added some exceptions to the loader deferral, because of the way that Label and Choice events are written they need to be loaded ahead of time or they won't build properly. I attached them to the events as a metadata, because creating that as a property directly in each one would take more time, but having it there would probably be better.

For further improvements for the next step, I think the best thing overall would be to take advantage of threading. As it is right now, every time it does one of these load or build operations, it's locking on the UI thread. First of all creating some kind of loading screen (as a scene that can be replaced so it can be customized) so the initial timeline load isn't locking up the game, and defer that to a thread until it's ready to go. then once it is playing the timeline, do a threaded load of the rest of the events, until they're all filled out for the rest of that timeline. and perhaps doing a read-ahead for say the next 5 or so lines, and if there is a jump to a new timeline coming up, start loading that timeline to prepare it to be ready to switch to.

Along with that (and part of the preloading the next timeline), perhaps another  function that can be called by the game that would _prepare_ but not _start_ a timeline until it's called, for if a game's spawning a timeline from some event in the game, such as talking to an NPC, but you don't know which one's going to be loaded, so then it can be preloaded along with say the rest of a level's load time and held until ready. That would take returning a prepared Timeline object and then using that for the start_timeline() and jumping straight into it to work, I think.
